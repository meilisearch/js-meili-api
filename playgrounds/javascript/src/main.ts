import "./style.css";
import { addDocuments, getAllHits, getSearchResponse } from "./meilisearch.js";

document.querySelector<HTMLDivElement>("#app")!.innerHTML = `
  <div>
    <h1>Meilisearch + Vite</h1>

    <h2>Documents:</h2>
    <div id="hits" style="white-space: break-spaces"> - </div>

    <h2>Search response:</h2>
    <div id="response" style="white-space: break-spaces"> - </div>

    <h2>Errors:</h2>
    <div id="errors">None</div>
  </div>
`;

function getErrorMessage(error: unknown): string {
  if (!(error instanceof Error)) {
    return JSON.stringify(error);
  }

  const message = String(error);

  if (error.cause === undefined) {
    return message;
  }

  return `${message}\nCaused by ${getErrorMessage(error.cause)}`;
}

try {
  await addDocuments();
  await getAllHits(document.querySelector<HTMLDivElement>("#hits")!);
  await getSearchResponse(document.querySelector<HTMLDivElement>("#response")!);
} catch (error) {
  console.error(error);
  document.querySelector<HTMLDivElement>("#errors")!.innerText =
    getErrorMessage(error);
}
