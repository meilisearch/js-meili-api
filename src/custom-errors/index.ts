import {
  AxiosError,
} from 'axios'
import MeiliSearchApiError from './meilisearch-error'
import MeilISearchCommunicationError from './meilisearch-communication-error'

function errorHandler (e: AxiosError, cachedStack?: string): void {
    if (e.response) {
      throw new MeiliSearchApiError(e, cachedStack)
    }
    else if (e.isAxiosError) {
      throw new MeilISearchCommunicationError(e.message)
    }
    else {
      throw e;
    }
  }

export { errorHandler }
