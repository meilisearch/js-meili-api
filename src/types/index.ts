export * from "./settings.js";
export * from "./types.js";
