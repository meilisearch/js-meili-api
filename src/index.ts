/*
 * Bundle: Meilisearch / Indexes
 * Project: Meilisearch - Javascript API
 * Author: Quentin de Quelen <quentin@meilisearch.com>
 * Copyright: 2019, Meilisearch
 */

'use strict'

import { MeiliSearchTimeOutError } from './errors/meilisearch-timeout-error'
import MeiliAxiosWrapper from './meili-axios-wrapper'
import * as Types from './types'
import { sleep } from './utils'

class Index extends MeiliAxiosWrapper implements Types.Index {
  uid: string
  constructor(config: Types.Config, uid: string) {
    super(config)
    this.uid = uid
  }

  ///
  /// UPDATES
  ///

  /**
   * Get the informations about an update status
   * @memberof Index
   * @method getUpdateStatus
   */
  getUpdateStatus(updateId: number): Promise<Types.Update> {
    const url = `/indexes/${this.uid}/updates/${updateId}`

    return this.get(url)
  }

  async waitForPendingUpdate(
    updateId: number,
    {
      timeOutMs = 5000,
      intervalMs = 50,
    }: { timeOutMs?: number; intervalMs?: number } = {}
  ) {
    const startingTime = Date.now()
    while (Date.now() - startingTime < timeOutMs) {
      const response = await this.getUpdateStatus(updateId)
      if (response.status !== 'enqueued') return response
      await sleep(intervalMs)
    }
    throw new MeiliSearchTimeOutError(
      `timeout of ${timeOutMs}ms has exceeded on process ${updateId} when waiting for pending update to resolve.`
    )
  }

  /**
   * Get the list of all updates
   * @memberof Index
   * @method getAllUpdateStatus
   */
  getAllUpdateStatus(): Promise<Types.Update[]> {
    const url = `/indexes/${this.uid}/updates`

    return this.get(url)
  }

  ///
  /// SEARCH
  ///

  /**
   * Search for documents into an index
   * @memberof Index
   * @method search
   */
  search(
    query: string,
    options?: Types.SearchParams
  ): Promise<Types.SearchResponse> {
    const url = `/indexes/${this.uid}/search`

    const params: Types.SearchRequest = {
      q: query,
    }
    if (options) {
      if (options.offset) {
        params.offset = options.offset
      }
      if (options.limit) {
        params.limit = options.limit
      }
      if (options.attributesToRetrieve) {
        if (Array.isArray(options.attributesToRetrieve)) {
          params.attributesToRetrieve = options.attributesToRetrieve.join(',')
        } else {
          params.attributesToRetrieve = options.attributesToRetrieve
        }
      }

      if (options.attributesToCrop) {
        if (Array.isArray(options.attributesToCrop)) {
          params.attributesToCrop = options.attributesToCrop.join(',')
        } else {
          params.attributesToCrop = options.attributesToCrop
        }
      }
      if (options.cropLength) {
        params.cropLength = options.cropLength
      }
      if (options.attributesToHighlight) {
        if (Array.isArray(options.attributesToHighlight)) {
          params.attributesToHighlight = options.attributesToHighlight.join(',')
        } else {
          params.attributesToHighlight = options.attributesToHighlight
        }
      }
      if (options.filters) {
        params.filters = options.filters
      }

      if (options.matches) {
        params.matches = options.matches
      }
    }

    return this.get(url, {
      params,
      cancelToken: this.cancelTokenSource.token,
    })
  }

  ///
  /// INDEX
  ///
  /**
   * Show index information.
   * @memberof Index
   * @method show
   */
  show(): Promise<Types.IndexResponse> {
    const url = `/indexes/${this.uid}`

    return this.get(url)
  }

  /**
   * Update an index.
   * @memberof Index
   * @method updateIndex
   */
  updateIndex(data: Types.UpdateIndexRequest): Promise<Types.IndexResponse> {
    const url = `/indexes/${this.uid}`

    return this.put(url, data)
  }

  /**
   * Delete an index.
   * @memberof Index
   * @method deleteIndex
   */

  deleteIndex(): Promise<string> {
    const url = `/indexes/${this.uid}`

    return this.delete(url)
  }

  ///
  /// STATS
  ///

  /**
   * get stats of an index
   * @memberof Index
   * @method getStats
   */
  getStats(): Promise<Types.IndexStats> {
    const url = `/indexes/${this.uid}/stats`

    return this.get(url)
  }
  ///
  /// DOCUMENTS
  ///

  /**
   * get documents of an index
   * @memberof Index
   * @method getDocuments
   */
  getDocuments(options?: Types.GetDocumentsParams): Promise<Types.Document[]> {
    const url = `/indexes/${this.uid}/documents`
    let attr
    if (options && Array.isArray(options.attributesToRetrieve)) {
      attr = options.attributesToRetrieve.join(',')
    }

    return this.get(url, {
      params: {
        ...options,
        ...(attr ? { attributesToRetrieve: attr } : {}),
      },
    })
  }

  /**
   * Get one document
   * @memberof Index
   * @method getDocument
   */
  getDocument(documentId: string | number): Promise<Types.Document> {
    const url = `/indexes/${this.uid}/documents/${documentId}`

    return this.get(url)
  }

  /**
   * Add or replace multiples documents to an index
   * @memberof Index
   * @method addDocuments
   */
  addDocuments(
    documents: Types.Document[],
    options?: Types.AddDocumentParams
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/documents`

    return this.post(url, documents, {
      params: options,
    })
  }

  /**
   * Add or update multiples documents to an index
   * @memberof Index
   * @method updateDocuments
   */
  updateDocuments(
    documents: Types.Document[],
    options?: Types.AddDocumentParams
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/documents`

    return this.put(url, documents, {
      params: options,
    })
  }

  /**
   * Delete one document
   * @memberof Index
   * @method deleteDocument
   */
  deleteDocument(documentId: string | number): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/documents/${documentId}`

    return this.delete(url)
  }

  /**
   * Delete multiples documents of an index
   * @memberof Index
   * @method deleteDocuments
   */
  deleteDocuments(
    documentsIds: string[] | number[]
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/documents/delete-batch`

    return this.post(url, documentsIds)
  }

  /**
   * Delete all documents of an index
   * @memberof Index
   * @method deleteAllDocuments
   */
  deleteAllDocuments(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/documents`

    return this.delete(url)
  }

  ///
  /// SETTINGS
  ///

  /**
   * Retrieve all settings
   * @memberof Index
   * @method getSettings
   */
  getSettings(): Promise<Types.Settings> {
    const url = `/indexes/${this.uid}/settings`

    return this.get(url)
  }

  /**
   * Update all settings
   * Any parameters not provided will be left unchanged.
   * @memberof Index
   * @method updateSettings
   */
  updateSettings(settings: Types.Settings): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings`

    return this.post(url, settings)
  }

  /**
   * Reset settings.
   * @memberof Index
   * @method resetSettings
   */
  resetSettings(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings`

    return this.delete(url)
  }

  ///
  /// SYNONYMS
  ///

  /**
   * Get the list of all synonyms
   * @memberof Index
   * @method getSynonyms
   */
  getSynonyms(): Promise<object> {
    const url = `/indexes/${this.uid}/settings/synonyms`

    return this.get(url)
  }

  /**
   * Update the list of synonyms. Overwrite the old list.
   * @memberof Index
   * @method updateSynonyms
   */
  updateSynonyms(synonyms: object): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/synonyms`

    return this.post(url, synonyms)
  }

  /**
   * Reset the synonym list to be empty again
   * @memberof Index
   * @method resetSynonyms
   */
  resetSynonyms(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/synonyms`

    return this.delete(url)
  }

  ///
  /// STOP WORDS
  ///

  /**
   * Get the list of all stop-words
   * @memberof Index
   * @method getStopWords
   */
  getStopWords(): Promise<string[]> {
    const url = `/indexes/${this.uid}/settings/stop-words`

    return this.get(url)
  }

  /**
   * Update the list of stop-words. Overwrite the old list.
   * @memberof Index
   * @method updateStopWords
   */
  updateStopWords(stopWords: string[]): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/stop-words`

    return this.post(url, stopWords)
  }

  /**
   * Reset the stop-words list to be empty again
   * @memberof Index
   * @method resetStopWords
   */
  resetStopWords(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/stop-words`

    return this.delete(url)
  }

  ///
  /// RANKING RULES
  ///

  /**
   * Get the list of all ranking-rules
   * @memberof Index
   * @method getRankingRules
   */
  getRankingRules(): Promise<string[]> {
    const url = `/indexes/${this.uid}/settings/ranking-rules`

    return this.get(url)
  }

  /**
   * Update the list of ranking-rules. Overwrite the old list.
   * @memberof Index
   * @method updateRankingRules
   */
  updateRankingRules(rankingRules: string[]): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/ranking-rules`

    return this.post(url, rankingRules)
  }

  /**
   * Reset the ranking rules list to its default value
   * @memberof Index
   * @method resetRankingRules
   */
  resetRankingRules(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/ranking-rules`

    return this.delete(url)
  }

  ///
  /// DISTINCT ATTRIBUTE
  ///

  /**
   * Get the distinct-attribute
   * @memberof Index
   * @method getDistinctAttribute
   */
  getDistinctAttribute(): Promise<string | void> {
    const url = `/indexes/${this.uid}/settings/distinct-attribute`

    return this.get(url)
  }

  /**
   * Update the distinct-attribute.
   * @memberof Index
   * @method updateDistinctAttribute
   */
  updateDistinctAttribute(
    distinctAttribute: string
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/distinct-attribute`

    return this.post(url, distinctAttribute)
  }

  /**
   * Reset the distinct-attribute.
   * @memberof Index
   * @method resetDistinctAttribute
   */
  resetDistinctAttribute(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/distinct-attribute`

    return this.delete(url)
  }

  ///
  /// SEARCHABLE ATTRIBUTE
  ///

  /**
   * Get the searchable-attributes
   * @memberof Index
   * @method getSearchableAttributes
   */
  getSearchableAttributes(): Promise<string[]> {
    const url = `/indexes/${this.uid}/settings/searchable-attributes`

    return this.get(url)
  }

  /**
   * Update the searchable-attributes.
   * @memberof Index
   * @method updateSearchableAttributes
   */
  updateSearchableAttributes(
    searchableAttributes: string[]
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/searchable-attributes`

    return this.post(url, searchableAttributes)
  }

  /**
   * Reset the searchable-attributes.
   * @memberof Index
   * @method resetSearchableAttributes
   */
  resetSearchableAttributes(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/searchable-attributes`

    return this.delete(url)
  }

  ///
  /// DISPLAYED ATTRIBUTE
  ///

  /**
   * Get the displayed-attributes
   * @memberof Index
   * @method getDisplayedAttributes
   */
  getDisplayedAttributes(): Promise<string[]> {
    const url = `/indexes/${this.uid}/settings/displayed-attributes`

    return this.get(url)
  }

  /**
   * Update the displayed-attributes.
   * @memberof Index
   * @method updateDisplayedAttributes
   */
  updateDisplayedAttributes(
    displayedAttributes: string[]
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/displayed-attributes`

    return this.post(url, displayedAttributes)
  }

  /**
   * Reset the displayed-attributes.
   * @memberof Index
   * @method resetDisplayedAttributes
   */
  resetDisplayedAttributes(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/displayed-attributes`

    return this.delete(url)
  }

  ///
  /// ACCEPT NEW FIELDS
  ///

  /**
   * Get the accept-new-fields value.
   * @memberof Index
   * @method getAcceptNewFields
   */
  getAcceptNewFields(): Promise<boolean> {
    const url = `/indexes/${this.uid}/settings/accept-new-fields`

    return this.get(url)
  }

  /**
   * Update the accept-new-fields value.
   * @memberof Index
   * @method updateAcceptNewFields
   */
  updateAcceptNewFields(
    acceptNewFields: boolean
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/accept-new-fields`

    return this.post(url, acceptNewFields)
  }
}

export { Index }
