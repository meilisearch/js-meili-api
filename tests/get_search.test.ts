import { ErrorStatusCode } from '../src/types'
import { EnqueuedTask } from '../src/enqueued-task'
import {
  clearAllIndexes,
  config,
  BAD_HOST,
  MeiliSearch,
  getClient,
  HOST,
  getKey,
} from './utils/meilisearch-test-utils'

const index = {
  uid: 'movies_test',
}
const emptyIndex = {
  uid: 'empty_test',
}

const dataset = [
  {
    id: 123,
    title: 'Pride and Prejudice',
    comment: 'A great book',
    genre: 'romance',
  },
  {
    id: 456,
    title: 'Le Petit Prince',
    comment: 'A french book about a prince that walks on little cute planets',
    genre: 'adventure',
  },
  {
    id: 2,
    title: 'Le Rouge et le Noir',
    comment: 'Another french book',
    genre: 'romance',
  },
  {
    id: 1,
    title: 'Alice In Wonderland',
    comment: 'A weird book',
    genre: 'adventure',
  },
  {
    id: 1344,
    title: 'The Hobbit',
    comment: 'An awesome book',
    genre: 'sci fi',
  },
  {
    id: 4,
    title: 'Harry Potter and the Half-Blood Prince',
    comment: 'The best book',
    genre: 'fantasy',
  },
  { id: 42, title: "The Hitchhiker's Guide to the Galaxy", genre: 'fantasy' },
]

jest.setTimeout(100 * 1000)

afterAll(() => {
  return clearAllIndexes(config)
})

describe.each([
  { permission: 'Master' },
  { permission: 'Admin' },
  { permission: 'Search' },
])('Test on GET search', ({ permission }) => {
  beforeAll(async () => {
    await clearAllIndexes(config)
    const client = await getClient('Master')
    const { taskUid: task1 } = await client.createIndex(index.uid)
    await client.waitForTask(task1)
    const { taskUid: task2 } = await client.createIndex(emptyIndex.uid)
    await client.waitForTask(task2)

    const newFilterableAttributes = ['genre', 'title', 'id']
    const { taskUid: task3 }: EnqueuedTask = await client
      .index(index.uid)
      .updateSettings({
        filterableAttributes: newFilterableAttributes,
        sortableAttributes: ['id'],
      })
    await client.waitForTask(task3)

    const { taskUid: task4 } = await client
      .index(index.uid)
      .addDocuments(dataset)
    await client.waitForTask(task4)
  })

  test(`${permission} key: Basic search`, async () => {
    const client = await getClient(permission)

    const response = await client.index(index.uid).searchGet('prince', {})

    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response).toHaveProperty('limit', 20)
    expect(response).toHaveProperty('offset', 0)
    expect(response).toHaveProperty('processingTimeMs', expect.any(Number))
    expect(response).toHaveProperty('query', 'prince')
    expect(response.hits.length).toEqual(2)
  })

  test(`${permission} key: search with options`, async () => {
    const client = await getClient(permission)
    const response = await client
      .index(index.uid)
      .searchGet('prince', { limit: 1 })
    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response).toHaveProperty('offset', 0)
    expect(response).toHaveProperty('limit', 1)
    expect(response).toHaveProperty('processingTimeMs', expect.any(Number))
    expect(response).toHaveProperty('query', 'prince')
    expect(response.hits.length).toEqual(1)
  })

  test(`${permission} key: search with sortable`, async () => {
    const client = await getClient(permission)
    const response = await client
      .index(index.uid)
      .search('', { sort: ['id:asc'] })
    expect(response).toHaveProperty('hits', expect.any(Array))
    const hit = response.hits[0]
    expect(hit.id).toEqual(1)
  })

  test(`${permission} key: search with array options`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('prince', {
      attributesToRetrieve: ['*'],
    })
    const hit = response.hits[0]

    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response).toHaveProperty('query', 'prince')
    expect(Object.keys(hit).join(',')).toEqual(
      Object.keys(dataset[1]).join(',')
    )
  })

  test(`${permission} key: search on attributesToSearchOn`, async () => {
    const client = await getClient(permission)

    const response = await client.index(index.uid).searchGet('prince', {
      attributesToSearchOn: ['id'],
    })

    expect(response.hits.length).toEqual(0)
  })

  test(`${permission} key: search on attributesToSearchOn set to null`, async () => {
    const client = await getClient(permission)

    const response = await client.index(index.uid).searchGet('prince', {
      attributesToSearchOn: null,
    })

    expect(response).toMatchSnapshot()
  })

  test(`${permission} key: search with options`, async () => {
    const client = await getClient(permission)
    const response = await client
      .index(index.uid)
      .searchGet('prince', { limit: 1 })
    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response).toHaveProperty('offset', 0)
    expect(response).toHaveProperty('limit', 1)
    expect(response).toHaveProperty('processingTimeMs', expect.any(Number))
    expect(response).toHaveProperty('query', 'prince')
    expect(response.hits.length).toEqual(1)
  })

  test(`${permission} key: search with limit and offset`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('prince', {
      limit: 1,
      offset: 1,
    })
    expect(response).toHaveProperty('hits', [
      {
        id: 4,
        title: 'Harry Potter and the Half-Blood Prince',
        comment: 'The best book',
        genre: 'fantasy',
      },
    ])
    expect(response).toHaveProperty('offset', 1)
    expect(response).toHaveProperty('limit', 1)
    expect(response).toHaveProperty('processingTimeMs', expect.any(Number))
    expect(response).toHaveProperty('query', 'prince')
    expect(response.hits.length).toEqual(1)
  })

  test(`${permission} key: search with matches parameter and small croplength`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('prince', {
      filter: 'title = "Le Petit Prince"',
      attributesToCrop: ['*'],
      cropLength: 5,
      showMatchesPosition: true,
    })
    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response.hits[0]).toHaveProperty('_matchesPosition', {
      comment: [{ start: 22, length: 6 }],
      title: [{ start: 9, length: 6 }],
    })
  })

  test(`${permission} key: search with all options but not all fields`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('prince', {
      limit: 5,
      offset: 0,
      attributesToRetrieve: ['id', 'title'],
      attributesToCrop: ['*'],
      cropLength: 6,
      attributesToHighlight: ['*'],
      filter: 'title = "Le Petit Prince"',
      showMatchesPosition: true,
    })

    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response).toHaveProperty('offset', 0)
    expect(response).toHaveProperty('limit', 5)
    expect(response).toHaveProperty('processingTimeMs', expect.any(Number))
    expect(response).toHaveProperty('query', 'prince')
    expect(response.hits[0]._formatted).toHaveProperty('title')
    expect(response.hits[0]._formatted).toHaveProperty('id')
    expect(response.hits[0]).not.toHaveProperty('comment')
    expect(response.hits[0]).not.toHaveProperty('description')
    expect(response.hits.length).toEqual(1)
    expect(response.hits[0]).toHaveProperty('_formatted', expect.any(Object))
    expect(response.hits[0]._formatted).toHaveProperty(
      'title',
      'Le Petit <em>Prince</em>'
    )
    expect(response.hits[0]).toHaveProperty(
      '_matchesPosition',
      expect.any(Object)
    )
  })

  test(`${permission} key: search on default cropping parameters`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('prince', {
      attributesToCrop: ['*'],
      cropLength: 6,
    })

    expect(response.hits[0]._formatted).toHaveProperty(
      'comment',
      '…book about a prince that walks…'
    )
  })

  test(`${permission} key: search on customized cropMarker`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('prince', {
      attributesToCrop: ['*'],
      cropLength: 6,
      cropMarker: '(ꈍᴗꈍ)',
    })

    expect(response.hits[0]._formatted).toHaveProperty(
      'comment',
      '(ꈍᴗꈍ)book about a prince that walks(ꈍᴗꈍ)'
    )
  })

  test(`${permission} key: search on customized highlight tags`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('prince', {
      attributesToHighlight: ['*'],
      highlightPreTag: '(⊃｡•́‿•̀｡)⊃ ',
      highlightPostTag: ' ⊂(´• ω •`⊂)',
    })

    expect(response.hits[0]._formatted).toHaveProperty(
      'comment',
      'A french book about a (⊃｡•́‿•̀｡)⊃ prince ⊂(´• ω •`⊂) that walks on little cute planets'
    )
  })

  test(`${permission} key: search with all options and all fields`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('prince', {
      limit: 5,
      offset: 0,
      attributesToRetrieve: ['*'],
      attributesToCrop: ['*'],
      cropLength: 6,
      attributesToHighlight: ['*'],
      filter: 'title = "Le Petit Prince"',
      showMatchesPosition: true,
    })
    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response).toHaveProperty('offset', 0)
    expect(response).toHaveProperty('limit', 5)
    expect(response).toHaveProperty('processingTimeMs', expect.any(Number))
    expect(response).toHaveProperty('query', 'prince')
    expect(response.hits.length).toEqual(1)
    expect(response.hits[0]).toHaveProperty('_formatted', expect.any(Object))
    expect(response.hits[0]._formatted).toHaveProperty(
      'title',
      'Le Petit <em>Prince</em>'
    )
    expect(response.hits[0]).toHaveProperty(
      '_matchesPosition',
      expect.any(Object)
    )
  })

  test(`${permission} key: search with all options but specific fields`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('prince', {
      limit: 5,
      offset: 0,
      attributesToRetrieve: ['id', 'title'],
      attributesToCrop: ['id', 'title'],
      cropLength: 6,
      attributesToHighlight: ['id', 'title'],
      filter: 'title = "Le Petit Prince"',
      showMatchesPosition: true,
    })
    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response).toHaveProperty('offset', 0)
    expect(response).toHaveProperty('limit', 5)
    expect(response).toHaveProperty('processingTimeMs', expect.any(Number))
    expect(response).toHaveProperty('query', 'prince')
    expect(response.hits.length).toEqual(1)
    expect(response.hits[0]).toHaveProperty('id', 456)
    expect(response.hits[0]).toHaveProperty('title', 'Le Petit Prince')
    expect(response.hits[0]).not.toHaveProperty('comment')
    expect(response.hits[0]).toHaveProperty('_formatted', expect.any(Object))
    expect(response.hits[0]).not.toHaveProperty(
      'description',
      expect.any(Object)
    )
    expect(response.hits[0]._formatted).toHaveProperty(
      'title',
      'Le Petit <em>Prince</em>'
    )
    expect(response.hits[0]._formatted).not.toHaveProperty('comment')
    expect(response.hits[0]).toHaveProperty(
      '_matchesPosition',
      expect.any(Object)
    )
  })

  test(`${permission} key: search with filter and facetDistribution`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('a', {
      filter: 'genre = romance',
      facets: ['genre'],
    })
    expect(response).toHaveProperty('facetDistribution', {
      genre: { romance: 2 },
    })
    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response.hits.length).toEqual(2)
  })

  test(`${permission} key: search with filter on number`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('a', {
      filter: 'id < 0',
      facets: ['genre'],
    })
    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response.hits.length).toEqual(0)
  })

  test(`${permission} key: search with filter with spaces`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('h', {
      filter: 'genre = "sci fi"',
    })
    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response.hits.length).toEqual(1)
  })

  test(`${permission} key: search with multiple filter`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('a', {
      filter: 'genre = romance AND (genre = romance OR genre = romance)',
      facets: ['genre'],
    })
    expect(response).toHaveProperty('facetDistribution', {
      genre: { romance: 2 },
    })
    expect(response).toHaveProperty('hits', expect.any(Array))
    expect(response.hits.length).toEqual(2)
  })

  test(`${permission} key: search with multiple filter and undefined query (placeholder)`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet(undefined, {
      filter: 'genre = fantasy',
      facets: ['genre'],
    })
    expect(response).toHaveProperty('facetDistribution', {
      genre: { fantasy: 2 },
    })
    expect(response.hits.length).toEqual(2)
  })

  test(`${permission} key: search with multiple filter and null query (placeholder)`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet(null, {
      filter: 'genre = fantasy',
      facets: ['genre'],
    })
    expect(response).toHaveProperty('facetDistribution', {
      genre: { fantasy: 2 },
    })
    expect(response.hits.length).toEqual(2)
    expect(response.estimatedTotalHits).toEqual(2)
  })

  test(`${permission} key: search with multiple filter and empty string query (placeholder)`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).searchGet('', {
      filter: 'genre = fantasy',
      facets: ['genre'],
    })

    expect(response).toHaveProperty('facetDistribution', {
      genre: { fantasy: 2 },
    })
    expect(response.hits.length).toEqual(2)
  })

  test(`${permission} key: Try to search with wrong format filter`, async () => {
    const client = await getClient(permission)
    await expect(
      client.index(index.uid).searchGet('prince', {
        filter: ['hello'],
      })
    ).rejects.toHaveProperty(
      'message',
      'The filter query parameter should be in string format when using searchGet'
    )
  })
  test(`${permission} key: search with vectors`, async () => {
    const client = await getClient(permission)
    const adminClient = await getClient('Admin')
    const adminKey = await getKey('Admin')

    await fetch(`${HOST}/experimental-features`, {
      body: JSON.stringify({ vectorStore: true }),
      headers: {
        Authorization: `Bearer ${adminKey}`,
        'Content-Type': 'application/json',
      },
      method: 'PATCH',
    })

    const { taskUid } = await adminClient
      .index(emptyIndex.uid)
      .updateEmbedders({
        default: {
          source: 'userProvided',
          dimensions: 1,
        },
      })
    await adminClient.waitForTask(taskUid)

    const response = await client
      .index(emptyIndex.uid)
      .searchGet('', { vector: [1], hybridSemanticRatio: 1.0 })

    expect(response).toHaveProperty('hits')
    expect(response).toHaveProperty('semanticHitCount')
    // Those fields are no longer returned by the search response
    // We want to ensure that they don't appear in it anymore
    expect(response).not.toHaveProperty('vector')
    expect(response).not.toHaveProperty('_semanticScore')
  })

  test(`${permission} key: search without vectors`, async () => {
    const client = await getClient(permission)
    const response = await client.index(index.uid).search('prince', {})

    expect(response).not.toHaveProperty('semanticHitCount')
  })

  test(`${permission} key: Try to search on deleted index and fail`, async () => {
    const client = await getClient(permission)
    const masterClient = await getClient('Master')
    const { taskUid } = await masterClient.index(index.uid).delete()
    await masterClient.waitForTask(taskUid)
    await expect(
      client.index(index.uid).searchGet('prince')
    ).rejects.toHaveProperty('code', ErrorStatusCode.INDEX_NOT_FOUND)
  })
})

describe.each([
  { host: BAD_HOST, trailing: false },
  { host: `${BAD_HOST}/api`, trailing: false },
  { host: `${BAD_HOST}/trailing/`, trailing: true },
])('Tests on url construction', ({ host, trailing }) => {
  test(`Test get search route`, async () => {
    const route = `indexes/${index.uid}/search`
    const client = new MeiliSearch({ host })
    const strippedHost = trailing ? host.slice(0, -1) : host
    await expect(client.index(index.uid).searchGet()).rejects.toHaveProperty(
      'message',
      `request to ${strippedHost}/${route} failed, reason: connect ECONNREFUSED ${BAD_HOST.replace(
        'http://',
        ''
      )}`
    )
  })

  test(`Test post search route`, async () => {
    const route = `indexes/${index.uid}/search`
    const client = new MeiliSearch({ host })
    const strippedHost = trailing ? host.slice(0, -1) : host
    await expect(client.index(index.uid).searchGet()).rejects.toHaveProperty(
      'message',
      `request to ${strippedHost}/${route} failed, reason: connect ECONNREFUSED ${BAD_HOST.replace(
        'http://',
        ''
      )}`
    )
  })
})
