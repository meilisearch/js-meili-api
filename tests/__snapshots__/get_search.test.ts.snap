// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on GET search Admin key: search on attributesToSearchOn set to null 1`] = `
{
  "estimatedTotalHits": 2,
  "hits": [
    {
      "comment": "A french book about a prince that walks on little cute planets",
      "genre": "adventure",
      "id": 456,
      "title": "Le Petit Prince",
    },
    {
      "comment": "The best book",
      "genre": "fantasy",
      "id": 4,
      "title": "Harry Potter and the Half-Blood Prince",
    },
  ],
  "limit": 20,
  "offset": 0,
  "processingTimeMs": 0,
  "query": "prince",
}
`;

exports[`Test on GET search Master key: search on attributesToSearchOn set to null 1`] = `
{
  "estimatedTotalHits": 2,
  "hits": [
    {
      "comment": "A french book about a prince that walks on little cute planets",
      "genre": "adventure",
      "id": 456,
      "title": "Le Petit Prince",
    },
    {
      "comment": "The best book",
      "genre": "fantasy",
      "id": 4,
      "title": "Harry Potter and the Half-Blood Prince",
    },
  ],
  "limit": 20,
  "offset": 0,
  "processingTimeMs": 0,
  "query": "prince",
}
`;

exports[`Test on GET search Search key: search on attributesToSearchOn set to null 1`] = `
{
  "estimatedTotalHits": 2,
  "hits": [
    {
      "comment": "A french book about a prince that walks on little cute planets",
      "genre": "adventure",
      "id": 456,
      "title": "Le Petit Prince",
    },
    {
      "comment": "The best book",
      "genre": "fantasy",
      "id": 4,
      "title": "Harry Potter and the Half-Blood Prince",
    },
  ],
  "limit": 20,
  "offset": 0,
  "processingTimeMs": 0,
  "query": "prince",
}
`;
